### PR TITLE
MCS-68 Integration Tests (Ready for Code Review)

### PR DESCRIPTION
**Pull request by ThomasSchellenbergNextCentury**
The goal of the integration tests is to verify that specific sequences of actions entered via the MCS Python API lead to specific metadata outputs containing expected observation, navigation, and object interaction results within tightly-controlled scenes.

A test is a single scene that's run using a specific list of actions at every metadata tier and validated against expected outputs for each tier. Each test is represented as a set of similarly-named files in the `integration_tests/data` folder. Please see the `integration_tests/README` for more details.

Suggestions on How to Review this PR:
- Review the python scripts in the `integration_tests` folder.
- Review the changes in the `machine_common_sense` folder.
- Look at a couple of the 250+ integration test data files in the `integration_tests/data` folder, to get an idea of how it works.
- Try running the integration tests (follow the instructions in `integration_tests/README`) using a version of Unity built from the latest code on the `development` ai2thor branch.

### CODE REVIEW

File: integration_tests/README.md [hunk 1]
```diff
@@ -0,0 +1,58 @@
+# MCS Integration Tests
+
+The goal of the integration tests is to:
+- Verify specific sequences of actions entered via the MCS Python API return specific metadata outputs containing expected observation, navigation, and object interaction results within tightly-controlled scenes (the "Handmade" tests).
+- Verify all objects that should be compiled into the given MCS Unity build actually exist in that build (the "Prefab" tests). Because the files for some objects (Unity Prefabs) are stored in a private repository and must be copied into the Resources folder before building, we want to verify that each object was correctly included.
+
+## Run Handmade Tests
+
+```
+python run_handmade_tests.py <mucs_unity_build_file_path>
```
Comment (rartiss55) [inline integration_tests/README.md:10]: Typo I think? <m**u**cs_unity_build_file_path>

File: integration_tests/run_tests.py [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+    # args = retrieve_test_args('All')
+    # start_prefab_tests(
+    #     args.mcs_unity_build_file_path,
+    #     args.mcs_unity_github_branch_name
+    # )
+    args = retrieve_test_args('All', handmade_only=True)
+    start_handmade_tests(
+        args.mcs_unity_build_file_path,
+        args.metadata,
+        args.test
```
Comment (rartiss55) [inline integration_tests/run_tests.py:24]: Is this missing the line "args.dev"?

**Comment by brianpippin** [comment]
Reviewing the code but wanted to get you this:  I got these errors when running flake8 on this branch: 

~/Documents/Projects/machineCommonSense/MCS[MCS-68-integration-tests] $ flake8
./integration_tests/run_tests.py:1:1: F401 'argparse' imported but unused
./integration_tests/run_tests.py:2:1: F401 'glob' imported but unused
./integration_tests/run_tests.py:3:1: F401 'json' imported but unused
./integration_tests/run_tests.py:4:1: F401 'math' imported but unused
./integration_tests/run_tests.py:5:1: F401 'os.path' imported but unused
./integration_tests/run_tests.py:7:1: F401 'machine_common_sense as mcs' imported but unused
./integration_tests/run_tests.py:8:1: F401 'integration_test_utils.METADATA_TIER_LIST' imported but unused
./integration_tests/run_tests.py:10:1: F401 'run_prefab_tests.start_prefab_tests' imported but unused
./integration_tests/run_prefab_tests.py:1:1: F401 'json' imported but unused
./integration_tests/run_prefab_tests.py:3:1: F401 'machine_common_sense as mcs' imported but unused
./integration_tests/run_prefab_tests.py:4:1: F401 'integration_test_utils.print_divider' imported but unused
./integration_tests/run_prefab_tests.py:15:80: E501 line too long (105 > 79 characters)

**Comment by rartiss55** [comment]
Good catch @brianpippin -- the new integration_tests/ folder might need to be added to the .pre-commit-config.yaml (or we might be able to remove the line checking specific directories).

**Comment by brianpippin** [comment]
So all the code looks good, however I got these errors running the tests, I am running ai2thor master build, should i have a different unity build for these?

FAILED INTEGRATION TESTS:
    005.ball_close_pickup (oracle): Step 1 failed:
        object testBall direction_y: -0.6 != -0.376377016
        object testBall direction_z: 0.8 != 0.9264666
        object testBall distance: 0.5 != 0.4317479
    006.ball_far_pickup (oracle): Step 3 failed:
        object testBall direction_y: -0.6 != -0.376377016
        object testBall direction_z: 0.8 != 0.9264666
        object testBall distance: 0.5 != 0.4317479
    007.ball_rotate_pickup (oracle): Step 13 failed:
        object testBall direction_x: 0.6928203 != 0.802343547
        object testBall direction_y: -0.6 != -0.376377016
        object testBall direction_z: 0.4000001 != 0.463233352
        object testBall distance: 0.5 != 0.4317479
    013.move_into_immobile_object (level1): Step 8 failed:
        return_status: SUCCESSFUL != OBSTRUCTED
    013.move_into_immobile_object (level2): Step 8 failed:
        return_status: SUCCESSFUL != OBSTRUCTED
    013.move_into_immobile_object (oracle): Step 8 failed:
        position_z: 0.8000001 != 0.7
        return_status: SUCCESSFUL != OBSTRUCTED
        object testBall direction_y: -0.4348922 != -0.3661855
        object testBall direction_z: 0.900482535 != 0.9305419
        object testBall distance: 0.488626838 != 0.580307
    015.pickup_unpickupable_object (level1): Step 1 failed:
        return_status: SUCCESSFUL != NOT_PICKUPABLE
    015.pickup_unpickupable_object (level2): Step 1 failed:
        return_status: SUCCESSFUL != NOT_PICKUPABLE
    015.pickup_unpickupable_object (oracle): Step 1 failed:
        return_status: SUCCESSFUL != NOT_PICKUPABLE
        object testBall direction_y: -0.6000001 != -0.207858756
        object testBall direction_z: 0.8 != 0.978158832
        object testBall distance: 0.5 != 1.02232885
        object testBall held: True != False
        object testBall position_z: 0.399999976 != 1.0
    016.pickup_multiple_object (oracle): Step 1 failed:
        object testBall1 direction_y: -0.6 != -0.376377016
        object testBall1 direction_z: 0.8 != 0.9264666
        object testBall1 distance: 0.5 != 0.4317479
    018.drop_large_object (oracle): Step 0 failed:
        object testSuitcase direction_x: 0.0203668941 != 0.0
        object testSuitcase direction_y: -0.142208487 != -0.3463501
        object testSuitcase direction_z: 0.9896271 != 0.9381053
        object testSuitcase position_x: 0.03457254 != 0.0
        object testSuitcase position_z: 1.67987943 != 1.25
    019.put_onto_non_receptacle (oracle): Step 1 failed:
        object testBall1 direction_y: -0.6 != -0.376377016
        object testBall1 direction_z: 0.8 != 0.9264666
        object testBall1 distance: 0.5 != 0.4317479
    020.put_onto_receptacle (oracle): Step 0 failed:
        object testSofa direction_y: 0.13402383 != -0.265156478
        object testSofa direction_z: 0.990978062 != 0.9642054
        object testSofa position_z: 1.50267744 != 1.5
        object testSofa texture_color_list: [] != ['grey']
    021.open_then_close_container (oracle): Step 0 failed:
        object testContainer direction_y: -0.07806342 != -0.419776082
        object testContainer direction_z: 0.9969483 != 0.9076277
    022.open_container_from_behind (oracle): Step 0 failed:
        object testContainer direction_y: -0.07802666 != -0.419776082
        object testContainer direction_z: 0.9969511 != 0.9076277
    023.open_and_close_non_container (level1): Step 1 failed:
        return_status: NOT_OPENABLE != NOT_RECEPTACLE
    023.open_and_close_non_container (level2): Step 1 failed:
        return_status: NOT_OPENABLE != NOT_RECEPTACLE
    023.open_and_close_non_container (oracle): Step 1 failed:
        return_status: NOT_OPENABLE != NOT_RECEPTACLE
    024.close_opened_container (oracle): Step 0 failed:
        object testContainer direction_y: 0.132734179 != -0.419776082
        object testContainer direction_z: 0.991151631 != 0.9076277
        object testContainer position_z: 1.29646957 != 1
    025.open_far_container (oracle): Step 0 failed:
        object testContainer direction_y: -0.0521262549 != -0.294644266
        object testContainer direction_z: 0.9986404 != 0.955607057
    026.close_far_container (oracle): Step 0 failed:
        object testContainer direction_y: 0.09619833 != -0.294644266
        object testContainer direction_z: 0.995362163 != 0.955607057
        object testContainer position_z: 1.79646564 != 1.5
    027.put_into_container (oracle): Step 0 failed:
        object testContainer direction_y: -0.06511042 != -0.359629
        object testContainer direction_z: 0.997878 != 0.9330954
    028.put_into_container_from_behind (oracle): Step 0 failed:
        object testContainer direction_y: -0.0650854 != -0.359629
        object testContainer direction_z: 0.9978796 != 0.9330954
    029.pickup_from_container (oracle): Step 0 failed:
        object testContainer direction_y: -0.129425481 != -0.6105057
        object testContainer direction_z: 0.9915889 != 0.7920119
    030.pickup_from_container_from_behind (oracle): Step 0 failed:
        object testContainer direction_y: -0.129327461 != -0.6105057
        object testContainer direction_z: 0.9916017 != 0.7920119
    031.stacking_blocks_bottom_to_top (oracle): Step 0 failed:
        object testBlock1 direction_x: 0.295396566 != 0.284247726
        object testBlock1 direction_y: -0.356938422 != -0.4382152
        object testBlock1 direction_z: 0.8861918 != 0.8527431
        object testBlock2 direction_x: -0.295398057 != -0.284247756
        object testBlock2 direction_y: -0.3569382 != -0.438215137
        object testBlock2 direction_z: 0.886191249 != 0.852743
    032.stacking_blocks_top_to_top (oracle): Step 0 failed:
        object testBlock1 direction_x: 0.295585334 != 0.304878265
        object testBlock1 direction_y: -0.356940866 != -0.266677529
        object testBlock1 direction_z: 0.8861278 != 0.914293349
        object testBlock2 direction_x: -0.295398057 != -0.284247756
        object testBlock2 direction_y: -0.3569382 != -0.438215137
        object testBlock2 direction_z: 0.886191249 != 0.852743
    033.stacking_blocks_bottom_to_bottom (oracle): Step 0 failed:
        object testBlock1 direction_x: 0.295396566 != 0.284247726
        object testBlock1 direction_y: -0.356938422 != -0.4382152
        object testBlock1 direction_z: 0.8861918 != 0.8527431
        object testBlock2 direction_x: -0.295148 != -0.3046899
        object testBlock2 direction_y: -0.35699138 != -0.2666944
        object testBlock2 direction_z: 0.886253238 != 0.9143512
    034.stacking_blocks_top_to_bottom (oracle): Step 0 failed:
        object testBlock1 direction_x: 0.295585334 != 0.304878265
        object testBlock1 direction_y: -0.356940866 != -0.266677529
        object testBlock1 direction_z: 0.8861278 != 0.914293349
        object testBlock2 direction_x: -0.295148 != -0.3046899
        object testBlock2 direction_y: -0.35699138 != -0.2666944
        object testBlock2 direction_z: 0.886253238 != 0.9143512
    035.stacking_blocks_side_to_top_v1 (oracle): Step 0 failed:
        object testBlock1 direction_x: 0.3746087 != 0.289999783
        object testBlock1 direction_y: -0.386315167 != -0.39874965
        object testBlock1 direction_z: 0.842869461 != 0.8699993
        object testBlock1 position_x: 0.400000036 != 0.3
        object testBlock2 direction_x: -0.295398057 != -0.284247756
        object testBlock2 direction_y: -0.3569382 != -0.438215137
        object testBlock2 direction_z: 0.886191249 != 0.852743
    036.stacking_blocks_side_to_top_v2 (oracle): Step 0 failed:
        object testBlock1 direction_x: 0.267244667 != 0.289999783
        object testBlock1 direction_y: -0.367461383 != -0.39874965
        object testBlock1 direction_z: 0.890815556 != 0.8699993
        object testBlock1 position_z: 1.0 != 0.9
        object testBlock2 direction_x: -0.295398057 != -0.284247756
        object testBlock2 direction_y: -0.3569382 != -0.438215137
        object testBlock2 direction_z: 0.886191249 != 0.852743
    037.stacking_blocks_bottom_to_side_v1 (oracle): Step 0 failed:
        object testBlock1 direction_x: 0.295396566 != 0.284247756
        object testBlock1 direction_y: -0.356938422 != -0.438215137
        object testBlock1 direction_z: 0.8861918 != 0.852743
        object testBlock2 direction_x: -0.3746087 != -0.289999783
        object testBlock2 direction_y: -0.386315167 != -0.39874965
        object testBlock2 direction_z: 0.842869461 != 0.8699993
        object testBlock2 position_x: -0.400000036 != -0.3
    038.stacking_blocks_bottom_to_side_v2 (oracle): Step 0 failed:
        object testBlock1 direction_x: 0.295396566 != 0.284247756
        object testBlock1 direction_y: -0.356938422 != -0.438215137
        object testBlock1 direction_z: 0.8861918 != 0.852743
        object testBlock2 direction_x: -0.316200346 != -0.289999783
        object testBlock2 direction_y: -0.434775442 != -0.39874965
        object testBlock2 direction_z: 0.8432008 != 0.8699993
        object testBlock2 position_z: 0.7999999 != 0.9
    039.stacking_blocks_many (oracle): Step 0 failed:
        object testBlock1 direction_x: 0.295396566 != 0.284247726
        object testBlock1 direction_y: -0.356938422 != -0.4382152
        object testBlock1 direction_z: 0.8861918 != 0.8527431
        object testBlock2 direction_x: 0.102757335 != 0.106175356
        object testBlock2 direction_y: -0.371658444 != -0.2784256
        object testBlock2 direction_z: 0.9226651 != 0.9545711
        object testBlock3 direction_x: -0.0920513645 != -0.10049583
        object testBlock3 direction_y: -0.379711866 != -0.414545327
        object testBlock3 direction_z: 0.9205137 != 0.904462636
        object testBlock3 position_z: 1.0 != 0.9
        object testBlock4 direction_x: -0.3746087 != -0.289999783
        object testBlock4 direction_y: -0.386315167 != -0.39874965
        object testBlock4 direction_z: 0.842869461 != 0.8699993
        object testBlock4 position_x: -0.400000036 != -0.3
    040.stacking_blocks_too_far (level1): Step 2 failed:
        return_status: SUCCESSFUL != OBSTRUCTED
    040.stacking_blocks_too_far (level2): Step 2 failed:
        return_status: SUCCESSFUL != OBSTRUCTED
    040.stacking_blocks_too_far (oracle): Step 0 failed:
        object testBlock1 direction_x: 0.271450251 != 0.262723029
        object testBlock1 direction_y: -0.3280032 != -0.405031234
        object testBlock1 direction_z: 0.9048363 != 0.87574327
        object testBlock2 direction_x: -0.271451652 != -0.262723029
        object testBlock2 direction_y: -0.328003049 != -0.405031234
        object testBlock2 direction_z: 0.904835939 != 0.87574327
    041.pickup_floor (level1): Step 1 failed:
        return_status: OBSTRUCTED != NOT_PICKUPABLE
    041.pickup_floor (level2): Step 1 failed:
        return_status: OBSTRUCTED != NOT_PICKUPABLE
    041.pickup_floor (oracle): Step 1 failed:
        return_status: OBSTRUCTED != NOT_PICKUPABLE
    042.put_onto_floor (level1): Step 2 failed:
        return_status: OUT_OF_REACH != NOT_RECEPTACLE
    042.put_onto_floor (level2): Step 2 failed:
        return_status: OUT_OF_REACH != NOT_RECEPTACLE
    042.put_onto_floor (oracle): Step 1 failed:
        object testBall direction_y: -0.6 != -0.376377016
        object testBall direction_z: 0.8 != 0.9264666
        object testBall distance: 0.5 != 0.4317479
    043.look_down (level1): Step 10 failed:
        head_tilt: 72 != 90
        return_status: SUCCESSFUL != CANNOT_ROTATE
    043.look_down (level2): Step 10 failed:
        head_tilt: 72 != 90
        return_status: SUCCESSFUL != CANNOT_ROTATE
    043.look_down (oracle): Step 10 failed:
        head_tilt: 72 != 90
        return_status: SUCCESSFUL != CANNOT_ROTATE
    044.look_up (level1): Step 10 failed:
        head_tilt: -72 != -90
        return_status: SUCCESSFUL != CANNOT_ROTATE
    044.look_up (level2): Step 10 failed:
        head_tilt: -72 != -90
        return_status: SUCCESSFUL != CANNOT_ROTATE
    044.look_up (oracle): Step 10 failed:
        head_tilt: -72 != -90
        return_status: SUCCESSFUL != CANNOT_ROTATE
    045.pickup_and_put_with_image_coord (oracle): Step 0 failed:
        object testSofa direction_y: 0.13402383 != -0.265156478
        object testSofa direction_z: 0.990978062 != 0.9642054
        object testSofa position_z: 1.50267744 != 1.5
        object testSofa texture_color_list: [] != ['grey']
    046.pickup_trophy_with_image_coord (oracle): Step 0 failed:
        object testTrophy direction_y: -0.321285218 != -0.423505276
        object testTrophy direction_z: 0.9469821 != 0.9058936
    047.open_and_close_with_image_coord (level1): Step 1 failed:
        return_status: OUT_OF_REACH != NOT_RECEPTACLE
    047.open_and_close_with_image_coord (level2): Step 1 failed:
        return_status: OUT_OF_REACH != NOT_RECEPTACLE
    047.open_and_close_with_image_coord (oracle): Step 0 failed:
        object testContainer direction_y: -0.07806277 != -0.419776082
        object testContainer direction_z: 0.9969484 != 0.9076277
    048.open_and_pickup_with_image_coord (oracle): Step 0 failed:
        object testContainer direction_y: -0.208131567 != -0.419776082
        object testContainer direction_z: 0.978100836 != 0.9076277
        object testTrophy direction_x: 0.0203890819 != 0.139115512
        object testTrophy direction_y: -0.349878132 != -0.3471392
        object testTrophy direction_z: 0.9365734 != 0.927438
        object testTrophy position_x: 0.0217890143 != 0.15
    049.pickup_then_walk (oracle): Step 1 failed:
        object testBall direction_y: -0.6 != -0.376377016
        object testBall direction_z: 0.8 != 0.9264666
        object testBall distance: 0.5 != 0.4317479
    050.drop_tall_object (level1): Step 2 failed:
        return_status: FAILED != SUCCESSFUL
    050.drop_tall_object (level2): Step 2 failed:
        return_status: FAILED != SUCCESSFUL
    050.drop_tall_object (oracle): Step 0 failed:
        object testTrophy position_x: 0.00439166324 != 0.0
    052.scripted_rotate (oracle): Step 6 failed:
        object testCube position_x: -0.20022355 != 0.0

**Comment by rartiss55** [comment]
> So all the code looks good, however I got these errors running the tests, I am running ai2thor master build, should i have a different unity build for these?
> 

You need to use the development branch.

**Comment by brianpippin** [comment]
> > So all the code looks good, however I got these errors running the tests, I am running ai2thor master build, should i have a different unity build for these?
> 
> @brianpippin You need to use the development branch.

Okay built ai2thor development, and running tests again.

**Comment by brianpippin** [comment]
Thanks @rartiss55 , now get:

FAILED INTEGRATION TESTS:
    NONE

**Review by brianpippin — APPROVED** [review]

**Comment by ThomasSchellenbergNextCentury** [comment]
Yay! Glad the tests work for both of you.

Thanks for pointing out the lint errors. I usually rely on the pre-commit script to notify me of lint errors. I didn't realize it wasn't running! It should be fixed now.

**Pull request merged**